text/0000-futures.md

```diff
@@ -0,0 +1,520 @@
+    /// be `poll`ed.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake` should place
+    /// the associated task onto this queue.
+    fn wake(self: &Arc<Self>);
+
+    /// Indicates that the associated task is ready to make progress and should be polled.
+    /// This function is like wake, but can only be called from the thread on which this
+    /// `Wake` was created.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake_local` should place
+    /// the associated task onto this queue.
+    unsafe fn wake_local(self: &Arc<Self>)
+}
+```
+
+To see how this might be used in practice, here's a simple example sketch:
+
+```rust
+struct ExecutorInner {
+    sync_ready_queue: SynchronousQueue,
+    optimized_queue: UnsafeCell<Vec<Task>>,
+}
+
+struct Task {
+    future: ...,
+    executor: Arc<ExecutorInner>,
+}
+
+impl Wake for Task {
+    fn wake(self: &Arc<Self>) {
+        self.executor.sync_ready_queue.push(self.clone());
+    }
+    unsafe fn wake_local(self: &Arc<Self>) {
+        (&mut *self.executor.optimized_queue.get()).push(self.clone())
+    }
+}
+```
+
+The use of `&Arc<Self>` rather than just `&self` makes it possible to work directly with
+the trait object for `Wake`, including cloning it. With `UnsafeWake` below, we'll see
+an API with greater flexibility for the cases where `Arc` is problematic.
+
+In general async values are not coupled to any particular executor, so we use trait
+objects to handle waking. These come in two forms: `Waker` for the general case, and
+`LocalWaker` to provide more effenciency when the wakeup is guaranteed to occur on the
+executor thread:
+
+```rust
+/// A `Waker` is a handle for waking up a task by notifying its executor that it
+/// is ready to be run.
+///
+/// This handle contains a trait object pointing to an instance of the `UnsafeWake`
+/// trait, allowing notifications to get routed through it.
+///
+/// Implements `Clone`, `Send`, and `Sync`.
+pub struct Waker { ... }
+
+impl Waker {
+    /// Wake up the task associated with this `Waker`.
+    pub fn wake(&self);
+}
+
+/// A `LocalWaker` is a handle for waking up a task by notifying its executor that it is ready to be run.
+///
+/// This is similar to the `Waker` type, but cannot be sent across threads. Task executors can use this type to implement more optimized singlethreaded wakeup behavior.
+impl LocalWaker {
+    /// Wake up the task associated with this `LocalWaker`.
+    pub fn wake(&self);
+}
+
+/// You can upgrade to a sendable `Waker` at zero cost, but waking through a `Waker` is more expensive
+/// due to synchronization.
+impl From<LocalWaker> for Waker  { .. }
+```
+
+Task execution always happens in the context of a `LocalWaker` that can be used to
+wake the task up locally, or converted into a `Waker` that can be sent to other threads.
+
+It's possible to construct a `Waker` using `From<Arc<dyn Wake>>`.
```

Comment:
Not right now, because `dyn Wake` is not `Wake`.

Reply:
Should be `Arc<impl Wake>` I suppose.